packages/forms/src/directives/validators.ts

```diff
@@ -69,6 +69,172 @@ export interface Validator {
+ *
+ */
+@Directive()
+abstract class AbstractValidatorDirective implements Validator, OnChanges {
+  private _validator: ValidatorFn = Validators.nullValidator;
+  private _onChange!: () => void;
+
+  /**
+   * Name of an input that matches directive selector attribute (e.g. `minlength` for
+   * `MinLengthDirective`). An input with a given name might contain configuration information (like
+   * `minlength='10'`) or a flag that indicates whether validator should be enabled (like
+   * `[required]='false'`).
+   *
+   * @nodoc
+   * @internal
+   */
+  abstract inputName: string;
+
+  /**
+   * Creates an instance of a validator (specific to a directive that extends this base class).
+   *
+   * @nodoc
+   * @internal
+   */
+  abstract createValidator(input: unknown): ValidatorFn;
+
+  /**
+   * Performs the necessary input normalization based on a specific logic of a Directive.
+   * For example, the function might be used to convert string-based representation of the
+   * `minlength` input to an integer value that can later be used in the `Validators.minLength`
+   * validator.
+   *
+   * @nodoc
+   * @internal
+   */
+  abstract normalizeInput(input: unknown): unknown;
+
+  /** @nodoc */
+  ngOnChanges(changes: SimpleChanges): void {
+    if (this.inputName in changes) {
+      const input = this.normalizeInput(changes[this.inputName].currentValue);
+      this._validator = this.createValidator(input);
+      if (this._onChange) {
+        this._onChange();
+      }
+    }
+  }
+
+  /** @nodoc */
+  validate(control: AbstractControl): ValidationErrors|null {
+    return this._validator(control);
+  }
+
+  /** @nodoc */
+  registerOnValidatorChange(fn: () => void): void {
+    this._onChange = fn;
+  }
+}
+
+/**
+ * @description
+ * Provider which adds `MaxValidator` to the `NG_VALIDATORS` multi-provider list.
+ */
+export const MAX_VALIDATOR: StaticProvider = {
+  provide: NG_VALIDATORS,
+  useExisting: forwardRef(() => MaxValidator),
+  multi: true
+};
+
+/**
+ * A directive which installs the {@link MaxValidator} for any `formControlName`,
+ * `formControl`, or control with `ngModel` that also has a `max` attribute.
+ *
+ * @see [Form Validation](guide/form-validation)
+ *
+ * @usageNotes
+ *
+ * ### Adding a max validator
+ *
+ * The following example shows how to add a max validator to an input attached to an
+ * ngModel binding.
+ *
+ * ```html
+ * <input type="number" ngModel max="4">
+ * ```
+ *
+ * @ngModule ReactiveFormsModule
+ * @ngModule FormsModule
+ * @publicApi
+ */
+@Directive({
+  selector:
+      'input[type=number][max][formControlName],input[type=number][max][formControl],input[type=number][max][ngModel]',
+  providers: [MAX_VALIDATOR],
+  host: {'[attr.max]': 'max ? max : null'}
+})
+export class MaxValidator extends AbstractValidatorDirective {
+  /**
+   * @description
+   * Tracks changes to the max bound to this directive.
+   */
+  @Input() max!: string;
+
+  inputName = 'max';
```

Comment:
We should add `@internal` annotation to MinValidator and MaxValidator classes as well, so that they don't show up in public API:

```suggestion
  /* @internal */
  inputName = 'max';
```
Actually it'd be even better to make them private:

```suggestion

  private inputName = 'max';
```

@sonukapoor could you please try to make them public?

Reply:
You mean private right?